File: output/hypersomnia/scripts/game/filters.lua
-- PHYSICS COLLISION LAYERS --
filters = {
	STATIC_OBJECT = "ALL",

	AVOIDANCE = {
		"STATIC_OBJECT"
	},
	
	DROPPED_ITEM = {
		"ITEM_PICK", "STATIC_OBJECT"
	},
	
	ITEM_PICK = {
		"DROPPED_ITEM"
	},
	
	CHARACTER = {
		"STATIC_OBJECT", "REMOTE_CHARACTER", "CHARACTER"
	},
	
	REMOTE_CHARACTER = {
		"BULLET", "CHARACTER", "SWING_HITSENSOR"
	},
	
	BULLET = {
		"STATIC_OBJECT", "REMOTE_CHARACTER"
	},
	
	REMOTE_BULLET = {
		"STATIC_OBJECT"
	},
	
	SWING_HITSENSOR = {
		"STATIC_OBJECT", "REMOTE_CHARACTER"
	}
}

-- processing

local all_keys = {}
local all_categories = {}

for k, v in pairs(filters) do
	table.insert(all_keys, k)
end

create_options(all_keys, all_categories)

for k, v in pairs(filters) do
	if type(v) == "string" and v == "ALL" then
		filters[k] = all_keys
	end
end

for k, v in pairs(filters) do
	local mask = 0
	
	for i=1, #filters[k] do
		mask = bitor(all_categories[filters[k][i]], mask)
	end
	
	filters[k] = create(b2Filter, {
		categoryBits = all_categories[k],
		maskBits = mask
	})
end

File: game_framework/globals/filters.cpp
#include "filters.h"

auto all = std::numeric_limits<decltype(b2Filter::categoryBits)>::max() & (~filters::TRIGGER);

namespace filters {
	b2Filter renderable() {
		b2Filter out;
		out.categoryBits = RENDERABLE;
		out.maskBits = RENDERABLE_QUERY;
		return out;
	}

	b2Filter renderable_query() {
		b2Filter out;
		out.categoryBits = RENDERABLE_QUERY;
		out.maskBits = RENDERABLE;
		return out;
	}
	
	b2Filter controlled_character() {
		b2Filter out;
		out.categoryBits = RENDERABLE | CONTROLLED_CHARACTER | DYNAMIC_OBJECT;
		out.maskBits = RENDERABLE_QUERY | CONTROLLED_CHARACTER | STATIC_OBJECT | DYNAMIC_OBJECT | REMOTE_CHARACTER | FRICTION_GROUND | TRIGGER;
		return out;
	}

	b2Filter dynamic_object() {
		b2Filter out;
		out.categoryBits = RENDERABLE | DYNAMIC_OBJECT;
		out.maskBits = RENDERABLE_QUERY | CONTROLLED_CHARACTER | STATIC_OBJECT | DYNAMIC_OBJECT | REMOTE_CHARACTER | FRICTION_GROUND;
		return out;
	}

	b2Filter static_object() {
		b2Filter out;
		out.categoryBits = RENDERABLE | STATIC_OBJECT;
		out.maskBits = RENDERABLE_QUERY | CONTROLLED_CHARACTER | DYNAMIC_OBJECT | REMOTE_CHARACTER | FRICTION_GROUND;
		return out;
	}

	b2Filter trigger() {
		b2Filter out;
		out.categoryBits = RENDERABLE | TRIGGER;
		out.maskBits = RENDERABLE_QUERY | TRIGGER | CONTROLLED_CHARACTER;
		return out;
	}
}